code/components/citizen-scripting-node/src/NodeScriptRuntime.cpp

```diff
@@ -0,0 +1,533 @@
+		SharedPushEnvironmentNoContext pushed(m_isolate);
+		
+		// create global state
+		v8::Local<v8::ObjectTemplate> global = v8::ObjectTemplate::New(m_isolate);
+
+		// create Citizen call routines
+		v8::Local<v8::ObjectTemplate> citizenObject = v8::ObjectTemplate::New(m_isolate);
+
+		// loop through routine list
+		for (auto& routine : g_citizenFunctions)
+		{
+			citizenObject->Set(m_isolate, routine.first.c_str(), v8::FunctionTemplate::New(m_isolate, routine.second, v8::External::New(m_isolate, this)));
+		}
+
+		// register the functions that are accessible from global scope
+		for (auto& routine : g_globalFunctions)
+		{
+			global->Set(m_isolate, routine.first.c_str(), v8::FunctionTemplate::New(m_isolate, routine.second, v8::External::New(m_isolate, this)));
+		}
+
+		// add a global 'print' function as alias for Citizen.trace for testing
+		global->Set(m_isolate, "print",	v8::FunctionTemplate::New(m_isolate, V8_Trace<NodeScriptRuntime, SharedPushEnvironmentNoContext>));
+
+		// set the Citizen object
+		global->Set(m_isolate, "Citizen", citizenObject);
+
+		// create a cfx alias too for the Citizen object
+		// global->Set(m_isolate, "cfx", citizenObject);
+
+		// create a V8 context and store it
+		const auto context = node::NewContext(m_isolate, global);
+		m_context.Reset(m_isolate, context);
+
+		// set the 'global' variable to the global itself (?) (NOTE: only works after context creation, maybe it shouldn't be done?)
+		global->Set(m_isolate, "global", global);
+
+		// store the ScRT in the context
+		context->SetEmbedderData(16, v8::External::New(m_isolate, this));
+
+		// allow "eval" and "new Function" (required by some fivem scripts)
+		context->AllowCodeGenerationFromStrings(true);
+
+		// context scope is required for environment creation
+		v8::Context::Scope scope(context);
+
+		// create and load our child node environment
+		m_nodeEnvironment = node::CreateEnvironment(m_isolateData, context, { m_resourceName }, {}, node::EnvironmentFlags::kNoCreateInspector);
+		node::LoadEnvironment(m_nodeEnvironment, g_envCode);
+	}
+	
+	// run system scripts
+	result_t hr;
+
+	// loading natives
+	if (FX_FAILED(hr = LoadSystemFile(const_cast<char*>(va("citizen:/scripting/v8/%s", "natives_server.js")))))
+	{
+		return hr;
+	}
+
+	// loading platform scripts
+	for (const char* platformScript : g_platformScripts)
+	{
+		if (FX_FAILED(hr = LoadSystemFile(const_cast<char*>(platformScript))))
+		{
+			return hr;
+		}
+	}
+	return FX_S_OK;
+}
+	
+result_t NodeScriptRuntime::Destroy()
+{
+	m_eventRoutine = TEventRoutine();
+	m_tickRoutine = std::function<void()>();
+	m_callRefRoutine = TCallRefRoutine();
+	m_deleteRefRoutine = TDeleteRefRoutine();
+	m_duplicateRefRoutine = TDuplicateRefRoutine();
+	
+	SharedPushEnvironment pushed(this);
+	
+	node::FreeEnvironment(m_nodeEnvironment);
+
+	m_context.Reset();
+	return FX_S_OK;
+}
+
+result_t NodeScriptRuntime::Tick()
+{
+	SharedPushEnvironment pushed(this);
+	
+	uv_run(m_uvLoop, UV_RUN_NOWAIT);
```

Comment:
Dropping the EnvironmentScope changeset here and using a nested libuv loop here seems risky - this would lead to any libuv callbacks being delayed by the server main tick rate, which could slow down _any_ nested IO operations by a _lot_. I'm hoping that you folks have deliberated this and are _certain_ you wish to go this way.

Reply:
@blattersturm
Indeed, thanks for pointing it out. I added a workaround for this case, however i'm not too happy about this solution. In general an optimal scenario is when workarounds are not needed, neither in the component nor within the nodejs src. Hopefully in the future we can work on this.

@thorium-cfx 
Node has its own libuv and v8 dependencies built with the binary, in order to have it easier from the embedder side, we have to rely on those instead of external ones like before, thats why its there.

Reply:
A "high frequency tick method" dedicated to pumping every Node resource's uv loop (even worse if it _also runs the main scheduler tick!_) is not the way to go here either. People at times run _hundreds_ of resources, and this runs counter to e.g. the tickless work done prior, even though this was meant to focus on _client_ performance, the same would apply to server-side overhead.

Tight coupling to reduce ScRT enumeration/tick overhead _still adds per-runtime cost_, makes a mess of the code, and still adds timer overhead/delay since event signaling _is not timer-bound at all_. There really isn't any way to cheap out of something like the EnvironmentScope stuff, _unless_ upstream nowadays has something like this via e.g. async hooks. If maintaining this patch set is problematic, consider suggesting this use case to upstream Node.js maintainers instead, given it's a common embedder use case to wish to embed in an existing libuv event loop (and the uv dependency isn't the painful one to wish to sync with upstream- it's the v8 one).

Perhaps a cheap trick could be to somehow hoist events from the Node.js event loop into the server event loop and wrap callbacks _that way_ rather than patching Node.js to enter/exit environments cleanly on callbacks.

(moved here from a top-level comment)

Reply:
For this approach, i'm entirely relying on past experience, where it just works fine with a high amount of resources, even though most gamemode logic was on server side. Since using this ScRT is optional, i would like to see if this punishes performance with real world examples while in the meantime we can work on alternative solutions like the ones you recommended.